src/libraries/System.Private.CoreLib/src/System/Globalization/CompareInfo.WebAssembly.cs

```diff
@@ -130,39 +130,75 @@ private unsafe int JsIndexOfCore(ReadOnlySpan<char> source, ReadOnlySpan<char> t
-        // we can skip them all (~1027k chars) by checking for the remaining UnicodeCategories (~291k chars)
+        // Control: 59 (out of 1105)
+        // Format: 43 (out of 731)
+        // NonSpacingMark: 195 (out of 18105)
+        // EnclosingMark: 5 (out of 221) // 0488, 0489, A670, A671, A672
+        // ModifierLetter: 2 (out of 4012) // 0640, 07FA
+        // SpacingCombiningMark: 4 (out of 4420) // 0F3E, 0F3F, 1CE1, 1CF7
+        // OtherPunctuation: 4 (out of 7004) // 180A, 1CD3, 10F86 (\uD803\uDC00), 10F87 (\uD803\uDF87)
+        // OtherLetter: 683 (out of 784142)
+        // OtherNotAssigned: 3581 (out of 24718)
+        // UppercaseLetter: 4 (out of 19159) // 10591 (\uD801\uDC91), 10592 (\uD801\uDC92), 10594 (\uD801\uDC94), 10595 (\uD801\uDC95)
+        // LowercaseLetter: 24 (out of 24565) // 10597 -  105AF
+        // OtherNumber: 1 (out of 5100) // 10FC6 (\uD843\uDFC6)
+        // PrivateUse: 614 (out of 108800)
+        // total: 5219 chars
         // skipping more characters than ICU would lead to hashes with smaller distribution and more collisions in hash tables
         // but it makes the behavior correct and consistent with locale-aware equals, which is acceptable tradeoff
-        private static bool ShouldNotBeSkipped(UnicodeCategory category) =>
-            category == UnicodeCategory.LowercaseLetter ||
-            category == UnicodeCategory.UppercaseLetter ||
-            category == UnicodeCategory.TitlecaseLetter ||
-            category == UnicodeCategory.LetterNumber ||
-            category == UnicodeCategory.OtherNumber ||
-            category == UnicodeCategory.Surrogate ||
-            category == UnicodeCategory.PrivateUse ||
-            category == UnicodeCategory.MathSymbol ||
-            category == UnicodeCategory.CurrencySymbol ||
-            category == UnicodeCategory.ModifierSymbol ||
-            category == UnicodeCategory.OtherSymbol ||
-            category == UnicodeCategory.OtherNotAssigned;
+        private static bool ShouldBeSkipped(UnicodeCategory category, char value)
+        {
+            switch (category)
+            {
+                case UnicodeCategory.Control:
+                case UnicodeCategory.Format:
+                case UnicodeCategory.NonSpacingMark:
+                case UnicodeCategory.OtherLetter:
+                case UnicodeCategory.OtherNotAssigned:
+                case UnicodeCategory.PrivateUse:
+                {
+                    return true;
+                }
+                case UnicodeCategory.LowercaseLetter:
+                {
+                    // some skipped unicodes, e.g. from Elbasan script, are surrogates
+                    int codePoint = char.ConvertToUtf32(value.ToString(), 0);
+                    return 0x10597 <= codePoint && codePoint <= 0x105AF;
+                }
+                case UnicodeCategory.UppercaseLetter:
+                {
+                    int codePoint = char.ConvertToUtf32(value.ToString(), 0);
```

Comment:
Do we need `char.ConvertToUtf32(value.ToString(), 0);` for all categories? I would just convert `value` to integer if we are sure that the category doesn't contain low/high surrogates.